### ISSUE TITLE
Question about fine-tuning on squad dataset

### ISSUE THREAD

**Issue by curtis0982** [google-research/electra#24]
I downloaded this model and I tried to use this model.
I choosed the squad 2.0 dataset to fine-tune.
When I tried to fine-tune the model on the command line,
the program just stopped working and the command line seemed to exit the program.
The output is like this:
```

(env_tf115) D:\python_code\NLP\electra>python run_finetuning.py --data-dir "D:\python_code\NLP\electra\datadir" --model-name electra_small --hparams {"model_size": "small", "task_names": ["squad"], "num_trials": 2}
2020-03-26 22:00:19.349133: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library cudart64_100.dll
{"model_size": "small", "task_names": ["squad"], "num_trials": 2}
================================================================================
Config: model=electra_small, trial 1/2
================================================================================
answerable_classifier True
answerable_uses_start_logits True
answerable_weight 0.5
beam_size 20
data_dir D:\python_code\NLP\electra\datadir
debug False
do_eval True
do_lower_case True
do_train True
doc_stride 128
double_unordered True
embedding_size 128
eval_batch_size 32
gcp_project None
init_checkpoint D:\python_code\NLP\electra\datadir\models\electra_small
iterations_per_loop 1000
joint_prediction True
keep_all_models True
layerwise_lr_decay 0.8
learning_rate 0.0001
log_examples False
max_answer_length 30
max_query_length 64
max_seq_length 512
model_dir D:\python_code\NLP\electra\datadir\models\electra_small\finetuning_models\squad_model
model_hparam_overrides {}
model_name electra_small
model_size small
n_best_size 20
n_writes_test 5
num_tpu_cores 1
num_train_epochs 2.0
num_trials 2
predict_batch_size 32
preprocessed_data_dir D:\python_code\NLP\electra\datadir\models\electra_small\finetuning_tfrecords\squad_tfrecords
qa_eval_file <built-in method format of str object at 0x0000027494EB70B0>
qa_na_file <built-in method format of str object at 0x0000027494EB1AE0>
qa_na_threshold -2.75
qa_preds_file <built-in method format of str object at 0x0000027494EB70F0>
raw_data_dir <built-in method format of str object at 0x0000027494EAECA8>
results_pkl D:\python_code\NLP\electra\datadir\models\electra_small\results\squad_results.pkl
results_txt D:\python_code\NLP\electra\datadir\models\electra_small\results\squad_results.txt
save_checkpoints_steps 1000000
task_names ['squad']
test_predictions <built-in method format of str object at 0x0000027494EAD8F0>
tpu_job_name None
tpu_name None
tpu_zone None
train_batch_size 32
use_tfrecords_if_existing True
use_tpu False
vocab_file D:\python_code\NLP\electra\datadir\models\electra_small\vocab.txt
vocab_size 30522
warmup_proportion 0.1
weight_decay_rate 0.01
write_distill_outputs False
write_test_outputs False

Loading dataset squad_train
Existing tfrecords not found so creating

(env_tf115) D:\python_code\NLP\electra>
```

My computer setting:
Windows 10
Cuda 10.0.130
Cudnn 7.6.3
Tensorflow 1.15 GPU

I've put the pre-trained model file under "datadir\models\electra_small" directory
and squad dataset under "datadir\finetuning_data\squad" directory.
Does anyone know why the fine-tuning is not working?

**Comment by clarkkev** [comment]
Huh, that is strange for it to exit with no error message. Can you add some print statements to find exactly at which point it exits?

**Comment by curtis0982** [comment]
I add some print statements in preprocessing.py to see what's going on:

`  def _serialize_dataset(self, tasks, is_training, split):
    """Write out the dataset as tfrecords."""
    dataset_name = "_".join(sorted([task.name for task in tasks]))
    dataset_name += "_" + split
    dataset_prefix = os.path.join(
        self._config.preprocessed_data_dir, dataset_name)
    tfrecords_path = dataset_prefix + ".tfrecord"
    metadata_path = dataset_prefix + ".metadata"
    batch_size = (self._config.train_batch_size if is_training else
                  self._config.eval_batch_size)

    utils.log("Loading dataset", dataset_name)
    n_examples = None
    if (self._config.use_tfrecords_if_existing and
        tf.io.gfile.exists(metadata_path)):
      n_examples = utils.load_json(metadata_path)["n_examples"]
    print("      n_examples = utils.load_json(metadata_path)[n_examples]\n")
    if n_examples is None:
      utils.log("Existing tfrecords not found so creating")
      examples = []
      print("\n examples = [] \n")
      for task in tasks:
        print("\ntask:",task,"\n")
        task_examples = task.get_examples(split)
        print("\n task_examples = task.get_examples(split)\n")
        examples += task_examples
        print("\nexamples += task_examples\n")
      if is_training:
        random.shuffle(examples)
        print("\n random.shuffle(examples) \n")
      utils.mkdir(tfrecords_path.rsplit("/", 1)[0])
      print("\n utils.mkdir(tfrecords_path.rsplit("/", 1)[0]) \n")
      n_examples = self.serialize_examples(
          examples, is_training, tfrecords_path, batch_size)
      print("\n n_examples = self.serialize_examples( \n")
      utils.write_json({"n_examples": n_examples}, metadata_path)

    input_fn = self._input_fn_builder(tfrecords_path, is_training)
    if is_training:
      steps = int(n_examples // batch_size * self._config.num_train_epochs)
    else:
      steps = n_examples // batch_size

    return input_fn, steps`

The output is like this:
`
Loading dataset squad_train
      n_examples = utils.load_json(metadata_path)[n_examples]

Existing tfrecords not found so creating
 examples = []
task: Task(squad)
(env_tf115) D:\python_code\NLP\electra>
`
It seems that what is not working properly is "task_examples = task.get_examples(split)" 
I'm trying to figure it out.

**Comment by curtis0982** [comment]
Updated:
In my case
Some code in finetune/qa/qa_tasks.py is not functioning properly.
'with tf.io.gfile.GFile(os.path.join(
        self.config.raw_data_dir(self.name),
        split + ("-debug" if self.config.debug else "") + ".json"), "r") as f:'
I printed
`os.path.join(self.config.raw_data_dir(self.name),split + ("-debug" if self.config.debug else "") + ".json` 
to see if the path is right.

My squad datadir located "D:\python_code\NLP\electra\datadir\finetuning_data\squad"
,so the squad training set path should be
"D:\python_code\NLP\electra\datadir\finetuning_data\squad\train.json"
But it is actually "squad\train.json"
So I changed this line 
`self.config.raw_data_dir(self.name)`
to my data_dir path.
The program started making tfrecord files, and then OOM and another question occurred.
Don't know if the errors are related to the original issue.

`  (0) Resource exhausted: OOM when allocating tensor with shape[32,4,512,512] and type float on /job:localhost/replica:0/task:0/device:GPU:0 by allocator GPU_0_bfc
         [[node electra/encoder/layer_7/attention/self/Softmax (defined at C:\WINDOWS\system32\env_tf115\lib\site-packages\tensorflow_core\python\framework\ops.py:1748) ]]
Hint: If you want to see a list of allocated tensors when OOM happens, add report_tensor_allocations_upon_oom to RunOptions for current allocation info.

         [[Sum/_4201]]
Hint: If you want to see a list of allocated tensors when OOM happens, add report_tensor_allocations_upon_oom to RunOptions for current allocation info.

  (1) Resource exhausted: OOM when allocating tensor with shape[32,4,512,512] and type float on /job:localhost/replica:0/task:0/device:GPU:0 by allocator GPU_0_bfc
         [[node electra/encoder/layer_7/attention/self/Softmax (defined at C:\WINDOWS\system32\env_tf115\lib\site-packages\tensorflow_core\python\framework\ops.py:1748) ]]
Hint: If you want to see a list of allocated tensors when OOM happens, add report_tensor_allocations_upon_oom to RunOptions for current allocation info.

0 successful operations.
0 derived errors ignored.

Original stack trace for 'electra/encoder/layer_7/attention/self/Softmax':
  File "run_finetuning.py", line 324, in <module>
    main()
  File "run_finetuning.py", line 320, in main
    args.model_name, args.data_dir, **hparams))
  File "run_finetuning.py", line 271, in run_finetuning
    model_runner.train()
  File "run_finetuning.py", line 184, in train
    input_fn=self._train_input_fn, max_steps=self.train_steps)
  File "C:\WINDOWS\system32\env_tf115\lib\site-packages\tensorflow_estimator\python\estimator\tpu\tpu_estimator.py", line 3030, in train
    saving_listeners=saving_listeners)
  File "C:\WINDOWS\system32\env_tf115\lib\site-packages\tensorflow_estimator\python\estimator\estimator.py", line 370, in train
    loss = self._train_model(input_fn, hooks, saving_listeners)
  File "C:\WINDOWS\system32\env_tf115\lib\site-packages\tensorflow_estimator\python\estimator\estimator.py", line 1161, in _train_model
    return self._train_model_default(input_fn, hooks, saving_listeners)
  File "C:\WINDOWS\system32\env_tf115\lib\site-packages\tensorflow_estimator\python\estimator\estimator.py", line 1191, in _train_model_default
    features, labels, ModeKeys.TRAIN, self.config)
  File "C:\WINDOWS\system32\env_tf115\lib\site-packages\tensorflow_estimator\python\estimator\tpu\tpu_estimator.py", line 2857, in _call_model_fn
    config)
  File "C:\WINDOWS\system32\env_tf115\lib\site-packages\tensorflow_estimator\python\estimator\estimator.py", line 1149, in _call_model_fn
    model_fn_results = self._model_fn(features=features, **kwargs)
  File "C:\WINDOWS\system32\env_tf115\lib\site-packages\tensorflow_estimator\python\estimator\tpu\tpu_estimator.py", line 3126, in _model_fn
    features, labels, is_export_mode=is_export_mode)
  File "C:\WINDOWS\system32\env_tf115\lib\site-packages\tensorflow_estimator\python\estimator\tpu\tpu_estimator.py", line 1663, in call_without_tpu
    return self._call_model_fn(features, labels, is_export_mode=is_export_mode)
  File "C:\WINDOWS\system32\env_tf115\lib\site-packages\tensorflow_estimator\python\estimator\tpu\tpu_estimator.py", line 1994, in _call_model_fn
    estimator_spec = self._model_fn(features=features, **kwargs)
  File "run_finetuning.py", line 86, in model_fn
    config, tasks, is_training, features, num_train_steps)
  File "run_finetuning.py", line 59, in __init__
    embedding_size=config.embedding_size)
  File "D:\python_code\NLP\electra\model\modeling.py", line 252, in __init__
    do_return_all_layers=True)
  File "D:\python_code\NLP\electra\model\modeling.py", line 870, in transformer_model
    to_seq_length=seq_length)
  File "D:\python_code\NLP\electra\model\modeling.py", line 747, in attention_layer
    attention_probs = tf.nn.softmax(attention_scores)
  File "C:\WINDOWS\system32\env_tf115\lib\site-packages\tensorflow_core\python\util\deprecation.py", line 507, in new_func
    return func(*args, **kwargs)
  File "C:\WINDOWS\system32\env_tf115\lib\site-packages\tensorflow_core\python\ops\nn_ops.py", line 2958, in softmax
    return _softmax(logits, gen_nn_ops.softmax, axis, name)
  File "C:\WINDOWS\system32\env_tf115\lib\site-packages\tensorflow_core\python\ops\nn_ops.py", line 2891, in _softmax
    return compute_op(logits, name=name)
  File "C:\WINDOWS\system32\env_tf115\lib\site-packages\tensorflow_core\python\ops\gen_nn_ops.py", line 11376, in softmax
    "Softmax", logits=logits, name=name)
  File "C:\WINDOWS\system32\env_tf115\lib\site-packages\tensorflow_core\python\framework\op_def_library.py", line 794, in _apply_op_helper
    op_def=op_def)
  File "C:\WINDOWS\system32\env_tf115\lib\site-packages\tensorflow_core\python\util\deprecation.py", line 507, in new_func
    return func(*args, **kwargs)
  File "C:\WINDOWS\system32\env_tf115\lib\site-packages\tensorflow_core\python\framework\ops.py", line 3357, in create_op
    attrs, op_def, compute_device)
  File "C:\WINDOWS\system32\env_tf115\lib\site-packages\tensorflow_core\python\framework\ops.py", line 3426, in _create_op_internal
    op_def=op_def)
  File "C:\WINDOWS\system32\env_tf115\lib\site-packages\tensorflow_core\python\framework\ops.py", line 1748, in __init__
    self._traceback = tf_stack.extract_stack()


(env_tf115) D:\python_code\NLP\electra>`
